### PR TITLE
Added solution for cf-1666F

### PR DESCRIPTION
**Pull request by Keys9873**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [_] I have tested my code.
- [_] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [_] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/gold/cf-1666F.mdx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+source: CF
+title: Fancy Stack
+author: Aakash Gokhale
+---
+
+[Official Editorial](https://codeforces.com/contest/1666/attachments/download/15881/nerc-2021-tutorial.pdf)
+
+## Explanation
+
+First we can solve the problem assuming all the blocks are distinct.
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:12]: what's the intuition behind this?

File: solutions/gold/cf-1666F.mdx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+Process the blocks from largest to smallest.
+
+We can do this problem using dp. Let $dp[i][j]$ be the number of ways to use the first $i$ blocks and place $j$ of those in the $j$ bottommost even positions (and $i - j$ of them in the odd position).
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:16]:
```suggestion
We can do this problem using dynamic programming.
Let $dp[i][j]$ be the number of ways to use the first $i$ blocks and place $j$ of those in the $j$ bottommost even positions (and $i - j$ of them in the odd position).
```

File: solutions/gold/cf-1666F.mdx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+There are two transitions.
+
+The first one will to $dp[i + 1][j + 1]$. If we add another even block, then there is only one place to put it.
+
+The second one will to $dp[i + 1][j]$. In this case we add the block to an odd position.
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:22]: use a numbered list

File: solutions/gold/cf-1666F.mdx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+Notice that we must place a block in between already placed larger blocks because of the condition that an even block is greater than its two neigboring odd blocks and later placed blocks will always be smaller.
+This means that there will be $max(0, j - 1)$ positions to insert the block. When all $\frac{n}{2}$ even blocks are placed you can also insert it at the top of the stack.
+We also have to subtract out the $i - j$ blocks that are already placed in odd positions.
+
+When the blocks can be of the same size, let's process them in groups of blocks that are all the same size
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:27]: missing period

File: solutions/gold/cf-1666F.mdx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+There are still two possible transitions.
+Either we put one block into an even position and the remaining into odd or all of them into odd positions.
+
+To determine the number of ways to insert the blocks into the odd positions we can do ${\textbf{number of open spots} \choose \textbf{number of spots that we want to fill}}$.
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:32]:
```suggestion
To determine the number of ways to insert the blocks into the odd positions, we can evaluate ${\textbf{number of open spots} \choose \textbf{number of spots that we want to fill}}$.
```

File: solutions/gold/cf-1666F.mdx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+Our answer will be $dp[n][\frac{n}{2}]$.
+
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:35]: add an implementation header

File: solutions/gold/cf-1666F.mdx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+using ll = long long;
+
+const ll mod = 998244353;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:47]: rename to MOD

File: solutions/gold/cf-1666F.mdx [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+const int MAXN = 5001;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:49]: use MAXN + 1 instead

File: solutions/gold/cf-1666F.mdx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	ll res = 1;
+	while (b) {
+		if (b & 1) res = (res * a) % mod;
+		a = (a * a) % mod;
+		b >>= 1;
+	}
+	return res;
+}
+
+void init() {
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:64]: just put this in main

File: solutions/gold/cf-1666F.mdx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+using namespace std;
+
+using ll = long long;
+
+const ll mod = 998244353;
+
+const int MAXN = 5001;
+ll f[MAXN];
+ll ivf[MAXN];
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:51]: since these are lookup tables, maybe capitalize these names asw?

File: solutions/gold/cf-1666F.mdx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+ll binpow(ll a, ll b) {
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:53]: link to the usaco guide link with this function
https://usaco.guide/gold/modular?lang=cpp

File: solutions/gold/cf-1666F.mdx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+	ll dp[n + 1][n + 1];
+	memset(dp, 0ll, sizeof(dp));
+	dp[0][0] = 1; 
+	ll size = 0;
+	for (int i = 0; i < ((int) b.size()); i++) {
+		for (ll j = 0; j <= size; j++) {
+			ll positions = max(j - 1, 0ll) + (j == n / 2) - (size - j);
+			// k in the even position and b[i].second - k in the odd
+			for (ll k = 0; k <= 1; k++) { 
+				if (j + k > n / 2) continue;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:103]: braces

File: solutions/gold/cf-1666F.mdx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+	cout << dp[n][n / 2] << endl;
+}
+
+int main() {
+	ios::sync_with_stdio(0);
+	cin.tie(0);
+	init();
+	int T = 1;
+	cin >> T;
+	while (T--) solve();
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:119]:
```suggestion
	int test_num;
	cin >> test_num;
	for (int t = 0; t < test_num; t++) {
		solve();
	}
```

File: solutions/gold/cf-1666F.mdx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+	f[0] = 1;
+	for (ll i = 1; i <= MAXN; i++) {
+		f[i] = (f[i - 1] * i) % mod;
+	}
+	for (ll i = 0; i <= MAXN; i++) {
+		ivf[i] = binpow(f[i], mod - 2);
+	}
+}
+
+ll c(ll a, ll b) {
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:74]: better name than a single character please

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
also instead of using an underscore to mark the checkboxes can you use an x? it actually fills them

File: solutions/gold/cf-1666F.mdx (hunk 1 above)
Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx]: Revised I think?

File: solutions/gold/cf-1666F.mdx [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+ 
+using namespace std;
+ 
+using ll = long long;
+ 
+const ll MOD = 998244353;
+ 
+const int MAXN = 5000;
+	
+// https://usaco.guide/gold/combo?lang=cpp
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:54]: where does the guide use `IVF` & stuff?

File: solutions/gold/cf-1666F.mdx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+	for (int i = 0; i < n; i++) {
+		if (b.size() > 0 && b.back().first == a[i]) {
+			b.back().second++;
+		} else {
+			b.push_back({a[i], 1ll});
+		}
+	}
+	// i greatest are used, j are even spots, i - j are odd spots
+	ll dp[n + 1][n + 1];
+	memset(dp, 0ll, sizeof(dp));
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:90]: just use a vector...

File: solutions/gold/cf-1666F.mdx [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+	for (int i = 0; i < ((int) b.size()); i++) {
+		for (ll j = 0; j <= size; j++) {
+			ll positions = max(j - 1, 0ll) + (j == n / 2) - (size - j);
+			// k in the even position and b[i].second - k in the odd
+			for (ll k = 0; k <= 1; k++) { 
+				if (j + k > n / 2) {
+					continue;
+				}
+				dp[size + b[i].second][j + k] = (dp[size + b[i].second][j + k] 
+					+ choose(positions, b[i].second - k) * dp[size][j] % MOD) % MOD;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:102]:
```suggestion
				dp[size + b[i].second][j + k] = (
					dp[size + b[i].second][j + k] 
					+ choose(positions, b[i].second - k)
					* dp[size][j] % MOD
				) % MOD;
```

File: solutions/gold/cf-1666F.mdx [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+			}
+		}
+		size += b[i].second;
+	}
+	cout << dp[n][n / 2] << endl;
+}
+ 
+int main() {
+	ios::sync_with_stdio(0);
+	cin.tie(0);
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:112]: does this run w/o?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/gold/cf-1666F.mdx (hunk 15 above)
Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx:54]: scroll down to method 2
i changed my code to the guide's code (it was basically the same)

Comment (SansPapyrus683) [reply solutions/gold/cf-1666F.mdx:54]: then can you change the names to align w/ the guide's as well?

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx:54]: i did now

File: solutions/gold/cf-1666F.mdx [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+author: Aakash Gokhale
+---
+
+[Official Editorial](https://codeforces.com/contest/1666/attachments/download/15881/nerc-2021-tutorial.pdf)
+
+## Explanation
+
+First we can solve an easier version of the problem where all the blocks are distinct.
+
+Process the blocks from largest to smallest.
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:14]: what does this have to do with anything

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx:14]: what do you mean?

Comment (SansPapyrus683) [reply solutions/gold/cf-1666F.mdx:14]: why do we process from big to small

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx]: Fixed i think? tried to make it more clear

File: solutions/gold/cf-1666F.mdx [hunk 20]
```diff
@@ -11,7 +11,7 @@ author: Aakash Gokhale
 
 First we can solve an easier version of the problem where all the blocks are distinct.
 
-Process the blocks from largest to smallest.
+Process the blocks from largest to smallest so we can incrementally fill the even positions starting from the bottom and working out way up to the top and also fill in the odd positions with whatever blocks we don't use in the even positions.
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:14]: seems like a runon sentence?

**Review by SansPapyrus683 — APPROVED** [review]
would like if you just submitted this code again to see if it works

File: solutions/gold/cf-1666F.mdx [hunk 21]
```diff
@@ -119,8 +119,8 @@ void solve() {
 				if (j + k > n / 2) { continue; }
 				dp[size + b[i].second][j + k] =
 				    (dp[size + b[i].second][j + k] +
-				     choose(positions, b[i].second - k, MOD) *
-				     dp[size][j] % MOD) %
+				     choose(positions, b[i].second - k, MOD) * dp[size][j] %
+				         MOD) %
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:123]: this is the WORST way they could've formatted this

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx]: I can't fix it because whenever I do the pre-commit bot just changes it back.

**Comment by Keys9873** [comment]
> would like if you just submitted this code again to see if it works

It does.

File: solutions/gold/cf-1666F.mdx (hunk 21 above)
Comment (SansPapyrus683) [reply solutions/gold/cf-1666F.mdx]: @bqi343 any idea why it formats like this?

Comment (bqi343) [reply solutions/gold/cf-1666F.mdx]: if you don't want it to look like this then split the statement into multiple statements

File: solutions/gold/cf-1666F.mdx [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+Notice that we must place a block in between already placed larger blocks because of the condition that an even block is greater than its two neigboring odd blocks and later placed blocks will always be smaller.
+This means that there will be $max(0, j - 1)$ positions to insert the block. When all $\frac{n}{2}$ even blocks are placed you can also insert it at the top of the stack.
+We also have to subtract out the $i - j$ blocks that are already placed in odd positions.
+
+When the blocks can be of the same size, let's process them in groups of blocks that are all the same size.
+
+There are still two possible transitions.
+Either we put one block into an even position and the remaining into odd or all of them into odd positions.
+
+To determine the number of ways to insert the blocks into the odd positions, we can evaluate ${\textbf{number of open spots} \choose \textbf{number of spots that we want to fill}}$.
```
Comment (bqi343) [inline solutions/gold/cf-1666F.mdx:34]: why not \text instead of \textbf

File: solutions/gold/cf-1666F.mdx [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+We can do this problem using dynamic programming.
+Let $dp[i][j]$ be the number of ways to use the first $i$ blocks and place $j$ of those in the $j$ bottommost even positions (and $i - j$ of them in the odd position).
+
+There are two transitions.
+
+1. To $dp[i + 1][j + 1]$. If we add another even block, then there is only one place to put it.
+
+2. To $dp[i + 1][j]$. In this case we add the block to an odd position.
+Notice that we must place a block in between already placed larger blocks because of the condition that an even block is greater than its two neigboring odd blocks and later placed blocks will always be smaller.
+This means that there will be $max(0, j - 1)$ positions to insert the block. When all $\frac{n}{2}$ even blocks are placed you can also insert it at the top of the stack.
```
Comment (bqi343) [inline solutions/gold/cf-1666F.mdx:26]:
```suggestion
This means that there will be $\max(0, j - 1)$ positions to insert the block. When all $\frac{n}{2}$ even blocks are placed you can also insert it at the top of the stack.
```

File: solutions/gold/cf-1666F.mdx [hunk 24]
```diff
@@ -117,11 +117,11 @@ void solve() {
...
-				dp[size + b[i].second][j + k] = (
-					dp[size + b[i].second][j + k] +
-					choose(positions, b[i].second - k, MOD) *
-					dp[size][j] % MOD
-				) % MOD;
+				dp[size + b[i].second][j + k] =
+				    (dp[size + b[i].second][j + k] +
+				     choose(positions, b[i].second - k, MOD) * dp[size][j] %
+				         MOD) %
+				    MOD;
```
Comment (SansPapyrus683) [inline solutions/gold/cf-1666F.mdx:124]: alright this is just straight unreasonable

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx]: I tried to make it less but now its making the indent smaller.
Im not sure how to increase the line limit (or if I can even do this from my end).

Comment (Keys9873) [reply solutions/gold/cf-1666F.mdx]: oh wait nvm i think it worked

**Review by ryanchou-dev — APPROVED** [review]
yessir
